### ISSUE TITLE
Error on final "@" in string?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#2510]
Migrated from [rt.perl.org#3973](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=3973) (status was 'resolved')

Searchable as RT3973$

**Comment by p5pRT** [comment]
### From daniel@lily.ebizquality.net

#### Created by daniel@chetlin.com



It's a bug in the core\, a bug in the docs\, or a bug in my understanding\.
&nbsp;  
&nbsp;  \[~\] $ perl57 \-le'print "foo $"'
&nbsp;  Final $ should be \\$ or $name at \-e line 1\, within string
&nbsp;  syntax error at \-e line 1\, near "print "foo $""
&nbsp;  Execution of \-e aborted due to compilation errors\.
&nbsp;  \[~\] $ perl57 \-le'print "foo @&#8203;"'
&nbsp;  foo @&#8203;

Whereas perldiag\.pod has the \_exact\_ same wording for both the final C\<$> and
the final C\<@&#8203;>\.
&nbsp;  
&nbsp;  =item Final $ should be \\$ or $name
&nbsp;  
&nbsp;  \(F\) You must now decide whether the final $ in a string was meant to be
&nbsp;  a literal dollar sign\, or was meant to introduce a variable name that
&nbsp;  happens to be missing\.  So you have to put either the backslash or the
&nbsp;  name\.
&nbsp;  
&nbsp;  =item Final @&#8203; should be \\@&#8203; or @&#8203;name
&nbsp;  
&nbsp;  \(F\) You must now decide whether the final @&#8203; in a string was meant to be
&nbsp;  a literal "at" sign\, or was meant to introduce a variable name that
&nbsp;  happens to be missing\.  So you have to put either the backslash or the
&nbsp;  name\.

Searching xray turned up nothing\, but that's not a big surprise for me\. xray
and I don't get along\.

This is likely related to the fact that we really don't treat '$' and '@&#8203;' the
same in doublequotish strings\. But again\, I don't know if that's intended or
just an artifact of having interpolated '$' but not '@&#8203;' in Perl 1\.

&nbsp;  \[~\] $ perl57 \-le'print "foo $ bar"'
&nbsp;  foo 
&nbsp;  \[~\] $ perl57 \-le'print "foo @&#8203; bar"'
&nbsp;  foo @&#8203; bar

So again\, my feeling is that either the docs or the behavior should change\,
since I was unable to find any mention in the docs of '$' and '@&#8203;' behaving
differently in interpolation \(aside from the warning that used to be a fatal
about an interpolated array which hasn't been seen yet\)\.

\(P\.S\. In trying to understand this\, I played with toke\.c\. My guess is that the
different behavior comes from lines 1308\-1321 in toke\.c \-\- regardless of
whether it will change or not\, can someone confirm that this is the case?\)

\-dlc

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.7.0:

Configured by daniel at Tue Aug 29 23:09:52 PDT 2000.

Summary of my perl5 (revision 5.0 version 7 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.2.16, archname=i686-linux
    uname='linux lily.ebizquality.net 2.2.16 #6 smp fri aug 18 13:49:47 pdt 2000 i686 unknown '
    config_args='-Dprefix=/home/daniel/dev/perl57 -Uinstallusrbinperl -DDEBUGGING -Dusedevel -ders'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define 
    use64bitint=undef use64bitall=undef uselongdouble=undef usesocks=undef
  Compiler:
    cc='cc', optimize='-O2', gccversion=egcs-2.91.66 19990314/Linux (egcs-1.1.2 release), gccosandvers=
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64'
    stdchar='char', d_stdstdio=define, usevfork=false
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL6907


@INC for perl v5.7.0:
    /disk/home/daniel/devel/PerlLib
    /home/daniel/dev/perl57/lib/5.7.0/i686-linux
    /home/daniel/dev/perl57/lib/5.7.0
    /home/daniel/dev/perl57/lib/site_perl/5.7.0/i686-linux
    /home/daniel/dev/perl57/lib/site_perl/5.7.0
    /home/daniel/dev/perl57/lib/site_perl
    .


Environment for perl v5.7.0:
    HOME=/disk/home/daniel
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:/usr/local/bin/:/usr/local/bin/scp:/usr/X11R6/bin:/usr/local/bin:/disk/home/daniel/bin
    PERL5LIB=/disk/home/daniel/devel/PerlLib
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @simoncozens

On Wed\, Aug 30\, 2000 at 02&#8203;:29&#8203;:07AM \-0700\, Daniel Chetlin wrote&#8203;:
>   \[~\] $ perl57 \-le'print "foo $ bar"'
>   foo 
>   \[~\] $ perl57 \-le'print "foo @&#8203; bar"'
>   foo @&#8203; bar
 
> \(P\.S\. In trying to understand this\, I played with toke\.c\. My guess is that the
> different behavior comes from lines 1308\-1321 in toke\.c \-\- regardless of
> whether it will change or not\, can someone confirm that this is the case?\)

The above behaviour is due to the tests in those lines\, yes\.
Don't think it should change\, though\. Should be fixed to allow @&#8203;\+ and @&#8203;\-
to be double\-quoted\, however\.

<details><summary>Inline Patch</summary>

```diff
--- /home/simon/patchbay/bleadperl/toke.c~	Wed Aug 30 03:30:18 2000
+++ /home/simon/patchbay/bleadperl/toke.c	Wed Aug 30 03:30:29 2000
@@ -1307,7 +1307,7 @@
 
 	/* check for embedded arrays (@foo, @:foo, @'foo, @{foo}, @$foo) */
 	else if (*s == '@' && s[1]
-		 && (isALNUM_lazy_if(s+1,UTF) || strchr(":'{$", s[1])))
+		 && (isALNUM_lazy_if(s+1,UTF) || strchr("+-:'{$", s[1])))
 	    break;
 
 	/* check for embedded scalars.  only stop if we're sure it's a
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

>The above behaviour is due to the tests in those lines\, yes\.
>Don't think it should change\, though\. Should be fixed to allow @&#8203;\+ and @&#8203;\-
>to be double\-quoted\, however\.

Yes; while working on the Camel\, Larry didn't let that \(ie "@&#8203;\+"
fails\) get documented\, because he said it was a bug that should get
fixed\.  \(When I'd brought it up before\, Ilya alleged that Larry
said that that was correct behaviour\, so I dropped it\, but I later
got direct confirmation that it was \*not\* correct\.\)

\-\-tom
